### PR TITLE
Pkcs11 - Support for Signing and Verification Functions for HMAC

### PR DESCRIPTION
**Pull request by ruchi393**

### CODE REVIEW

File: ta/pkcs11/src/processing_symm.c [hunk 1]
```diff
@@ -273,6 +311,41 @@ static enum pkcs11_rc input_data_size_is_valid(struct active_processing *proc,
...
+		break;
+	case PKCS11_CKM_SHA512_HMAC:
+		sign_sz = TEE_SHA512_HASH_SIZE;
+		break;
+	default:
+		return PKCS11_CKR_OK;
+	}
+
+	if (in_size < sign_sz)
+		return PKCS11_CKR_SIGNATURE_LEN_RANGE;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:344]: Could you add `PKCS11_CKR_SIGNATURE_LEN_RANGE` to `string_rc[]` in **pkcs11_helpers.c** now that it's used?

File: ta/pkcs11/include/pkcs11_ta.h [hunk 2]
```diff
@@ -341,6 +341,84 @@ enum pkcs11_ta_cmd {
...
+	PKCS11_CMD_VERIFY_UPDATE = 139,
+
+	/*
+	 * PKCS11_CMD_SIGN_FINAL - Finalize a signature computation processing
+	 *
+	 * [in]  memref[0] = 32bit session handle
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [out] memref[2] = output signature
+	 *
+	 * These commands relate to the PKCS#11 API functions C_SignFinal()
```
Comment (etienne-lms) [inline ta/pkcs11/include/pkcs11_ta.h:383]: Typo: "This command relates..."

File: ta/pkcs11/include/pkcs11_ta.h [hunk 3]
```diff
@@ -341,6 +341,84 @@ enum pkcs11_ta_cmd {
...
+	PKCS11_CMD_SIGN_FINAL = 140,
+
+	/*
+	 * PKCS11_CMD_VERIFY_FINAL - Finalize a signature verification processing
+	 *
+	 * [in]  memref[0] = 32bit session handle
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [in]  memref[2] = input signature to be processed
+	 *
+	 * These commands relate to the PKCS#11 API functions C_VerifyFinal()
```
Comment (etienne-lms) [inline ta/pkcs11/include/pkcs11_ta.h:394]: Ditto

**Comment by etienne-lms** [comment]
I found some minor stuff (see review comments) + checkpatch (Travis/CI) complainng on lines exceeding 80 chars.

File: ta/pkcs11/src/processing_symm.c (hunk 1 above)
Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c:344]: Done

File: ta/pkcs11/include/pkcs11_ta.h [hunk 4]
```diff
@@ -341,6 +341,84 @@ enum pkcs11_ta_cmd {
...
+	 *              32bit key handle,
+	 *              (struct pkcs11_attribute_head)mechanism + mecha params,
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 *
+	 * These commands relate to the PKCS#11 API functions C_SignInit() and
+	 * C_VerifyInit().
+	 */
+	PKCS11_CMD_SIGN_INIT = 136,
+	PKCS11_CMD_VERIFY_INIT = 137,
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply ta/pkcs11/include/pkcs11_ta.h]: Done

File: ta/pkcs11/include/pkcs11_ta.h (hunk 2 above)
Comment (ruchi393) [reply ta/pkcs11/include/pkcs11_ta.h]: Changed

File: ta/pkcs11/include/pkcs11_ta.h (hunk 3 above)
Comment (ruchi393) [reply ta/pkcs11/include/pkcs11_ta.h]: Changed

File: ta/pkcs11/src/entry.c [hunk 5]
```diff
@@ -247,7 +247,44 @@ TEE_Result TA_InvokeCommandEntryPoint(void *tee_session, uint32_t cmd,
...
+		rc = entry_processing_init(client, ptypes, params,
+					   PKCS11_FUNCTION_VERIFY);
+		break;
+	case PKCS11_CMD_SIGN_ONESHOT:
+		rc = entry_processing_step(client, ptypes, params,
+					   PKCS11_FUNCTION_SIGN,
+					   PKCS11_FUNC_STEP_ONESHOT);
+		break;
+	case PKCS11_CMD_VERIFY_ONESHOT:
+		rc = entry_processing_step(client, ptypes, params, // IN+IN
```
Comment (etienne-lms) [inline ta/pkcs11/src/entry.c:264]: maybe remove  `// IN+IN`.

File: ta/pkcs11/src/processing_symm.c [hunk 6]
```diff
@@ -339,6 +412,33 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 	 * (PKCS11_FUNC_STEP_UPDATE/_ONESHOT)
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:412]: Need to update comment : inconsistent with lines 421/422 :(
I suggest to remove line 339 and only keep line 338 ` * Feed active operation with data`.

**Comment by ruchi393** [comment]
Checkpatch error for line exceeding 80 chars should go away once I squash the patches.

File: ta/pkcs11/src/processing_symm.c (hunk 6 above)
Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c]: Done

File: ta/pkcs11/src/entry.c (hunk 5 above)
Comment (ruchi393) [reply ta/pkcs11/src/entry.c]: Done

**Comment by jforissier** [comment]
> Checkpatch error for line exceeding 80 chars should go away once I squash the patches.

Correct, as you can see in the Travis log the checkpatch run on the combined diff doesn't show the error anymore.

Please fix the "space before tabs" warnings though.

**Comment by ruchi393** [comment]
I have squashed the commits and also fixed the "space before tabs" warning.

File: ta/pkcs11/src/processing.c [hunk 7]
```diff
@@ -276,65 +283,3 @@ enum pkcs11_rc entry_processing_step(struct pkcs11_client *client,
...
-	DMSG("PKCS11 session %"PRIu32": verify %s %s: %s", session->handle,
-	     id2str_proc(mecha_type), id2str_function(function),
-	     id2str_rc(rc));
-
-out:
-	if (rc != PKCS11_CKR_BUFFER_TOO_SMALL)
-		release_active_processing(session);
-
-	return rc;
-}
```
Comment (jforissier) [inline ta/pkcs11/src/processing.c:340]: If this function is not needed, it should be removed in a separate commit.

Comment (ruchi393) [reply ta/pkcs11/src/processing.c:340]: Ok..I have created a separate patch for this.

**Review by etienne-lms — COMMENTED** [review]
commit: "ta: pkcs11: Remove unused function entry_verify_oneshot()":
Remove also `entry_verify_oneshot()` from **ta/pkcs11/src/processing.h**.

File: ta/pkcs11/include/pkcs11_ta.h [hunk 8]
```diff
@@ -341,6 +341,90 @@ enum pkcs11_ta_cmd {
...
+	PKCS11_CMD_VERIFY_UPDATE = 28,
+
+	/*
+	 * PKCS11_CMD_SIGN_FINAL - Finalize a signature computation processing
+	 *
+	 * [in]  memref[0] = 32bit session handle
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [out] memref[2] = output signature
+	 *
+	 * This command relates to the PKCS#11 API functions C_SignFinal()
```
Comment (etienne-lms) [inline ta/pkcs11/include/pkcs11_ta.h:386]: nitpicking: terminal dot here and at line 398.

File: ta/pkcs11/include/pkcs11_ta.h [hunk 9]
```diff
@@ -341,6 +341,90 @@ enum pkcs11_ta_cmd {
...
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [in]  memref[2] = input signature to be processed
+	 *
+	 * This command relates to the PKCS#11 API functions C_VerifyFinal()
+	 */
+	PKCS11_CMD_VERIFY_FINAL = 30,
+
+	/*
+	 * PKCS11_CMD_SIGN_ONESHOT - Update and finalize a signature
+	 *                           computation
```
Comment (etienne-lms) [inline ta/pkcs11/include/pkcs11_ta.h:404]: I would change `Update and finalize a signature computation` to `Compute a signature`.
Ditto at line 416. Highlights it's not a update+final, it is a full one-step computation.

File: ta/pkcs11/src/processing_symm.c [hunk 10]
```diff
@@ -273,6 +311,41 @@ static enum pkcs11_rc input_data_size_is_valid(struct active_processing *proc,
...
+		sign_sz = TEE_SHA256_HASH_SIZE;
+		break;
+	case PKCS11_CKM_SHA384_HMAC:
+		sign_sz = TEE_SHA384_HASH_SIZE;
+		break;
+	case PKCS11_CKM_SHA512_HMAC:
+		sign_sz = TEE_SHA512_HASH_SIZE;
+		break;
+	default:
+		return PKCS11_CKR_OK;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:340]: I think `PKCS11_CKR_ERROR_GENERIC` would be better here.

**Review by etienne-lms — COMMENTED** [review]
some nitpicking comments.

**Comment by ruchi393** [comment]
@jforissier , @etienne-lms , should I squash the commits ?

**Comment by ruchi393** [comment]
@Emantor , @jenswi-linaro  can you also please review. 
There are corresponding PR's in optee_client and optee_test repos also.
https://github.com/OP-TEE/optee_client/pull/232
https://github.com/OP-TEE/optee_test/pull/461

**Comment by etienne-lms** [comment]
@ricardosalveti, maybe you can help reviewing this.
@ruchi393, I think yes you can squash now to ease review.

**Comment by vesajaaskelainen** [comment]
Note: I have more fixes for HMAC keys. But that triggers from copying keys so wondering if they should be fixed in this PR or shall I submit it later?

File: ta/pkcs11/src/processing_symm.c [hunk 11]
```diff
@@ -336,9 +409,35 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 
 	/*
 	 * Feed active operation with data
-	 * (PKCS11_FUNC_STEP_UPDATE/_ONESHOT)
```
Comment (ricardosalveti) [inline ta/pkcs11/src/processing_symm.c:339]: Why removing this comment here but not later when finalizing?

**Comment by ricardosalveti** [comment]
> Note: I have more fixes for HMAC keys. But that triggers from copying keys so wondering if they should be fixed in this PR or shall I submit it later?

What are the issues involved with the fixes you have? If not specifically related to this PR, probably better to create a new one.

File: ta/pkcs11/src/processing_symm.c (hunk 11 above)
Comment (ruchi393) [reply ta/pkcs11/src/processing_symm.c:339]: For _ONESHOT Operation, for HMAC and AES cases, no active operation is performed.  It's only the update process where active operation is fed with data. Hence, we removed the comment from here and kept it as just "Feed active operation with data"

**Comment by ruchi393** [comment]
> @ricardosalveti, maybe you can help reviewing this.
> @ruchi393, I think yes you can squash now to ease review.

All review comments till now have been addresses, patches squashed and branch pushed for review

**Comment by vesajaaskelainen** [comment]
> > Note: I have more fixes for HMAC keys. But that triggers from copying keys so wondering if they should be fixed in this PR or shall I submit it later?
> 
> What are the issues involved with the fixes you have? If not specifically related to this PR, probably better to create a new one.

I am currently working with NXP optee-based tree which I cannot share at the moment and cherry-picking from there wasn't too easy :|... I try to get it cleaned up and make branch on top of optee masters so that I can push all my pkcs11 changes that I have for others to see too.

File: ta/pkcs11/src/processing_symm.c (hunk 11 above)
Comment (ricardosalveti) [reply ta/pkcs11/src/processing_symm.c:339]: That's fine, thanks for explaining it.

**Comment by ricardosalveti** [comment]
> > > Note: I have more fixes for HMAC keys. But that triggers from copying keys so wondering if they should be fixed in this PR or shall I submit it later?
> > 
> > 
> > What are the issues involved with the fixes you have? If not specifically related to this PR, probably better to create a new one.
> 
> I am currently working with NXP optee-based tree which I cannot share at the moment and cherry-picking from there wasn't too easy :|... I try to get it cleaned up and make branch on top of optee masters so that I can push all my pkcs11 changes that I have for others to see too.

I would suggest to create a new PR with the fixes then, as it seems it won't be trivial to get them here.

`Reviewed-by: Ricardo Salveti <ricardo@foundries.io>`

Will take a look at the client and test prs now.

File: ta/pkcs11/src/token_capabilities.c [hunk 12]
```diff
@@ -225,6 +238,30 @@ void mechanism_supported_key_sizes(uint32_t proc_id, uint32_t *min_key_size,
 				   uint32_t *max_key_size)
 {
 	switch (proc_id) {
+	case PKCS11_CKM_MD5_HMAC:
+		*min_key_size = 16;
```
Comment (vesajaaskelainen) [inline ta/pkcs11/src/token_capabilities.c:242]: Why not use TEE_MD5_HASH_SIZE and friends in here like what is on input_sign_size_is_valid()?

Comment (ruchi393) [reply ta/pkcs11/src/token_capabilities.c:242]: > Why not use TEE_MD5_HASH_SIZE and friends in here like what is on input_sign_size_is_valid()?

I can do that but the min and max key sizes over here are not accurate which I am planning to replace by another PR. I left them here to be consistent with the sks code from @etienne-lms . The Global Platform Internal API used for HMAC support different min and max key sizes not limited to hash digest size.  Once this PR is merged, I plan to submit another PR for the same.

**Comment by ruchi393** [comment]
Rebased and tags applied.
